nifi-nar-bundles/nifi-py4j-bundle/nifi-python-framework/src/main/python/framework/ExtensionManager.py

```diff
@@ -0,0 +1,531 @@
+
+    def getDependencies(self):
+        list = self.gateway.jvm.java.util.ArrayList()
+        for dep in self.dependencies:
+            list.add(dep)
+
+        return list
+
+    def getCapabilityDescription(self):
+        return self.description
+
+    def getTags(self):
+        list = self.gateway.jvm.java.util.ArrayList()
+        for tag in self.tags:
+            list.add(tag)
+
+        return list
+
+
+
+
+class ExtensionManager:
+    """
+    ExtensionManager is responsible for discovery of extensions types and the lifecycle management of those extension types.
+    Discovery of extension types includes finding what extension types are available
+    (e.g., which Processor types exist on the system), as well as information about those extension types, such as
+    the extension's documentation (tags and capability description).
+
+    Lifecycle management includes determining the third-party dependencies that an extension has and ensuring that those
+    third-party dependencies have been imported.
+    """
+
+    processorInterfaces = ['org.apache.nifi.python.processor.FlowFileTransform', 'org.apache.nifi.python.processor.RecordTransform']
+    processor_details = {}
+    processor_class_by_name = {}
+    module_files_by_extension_type = {}
+    dependency_directories = {}
+
+    def __init__(self, gateway):
+        self.gateway = gateway
+
+
+    def getProcessorTypes(self):
+        """
+        :return: a list of Processor types that have been discovered by the #discoverExtensions method
+        """
+        return self.processor_details.values()
+
+    def getProcessorClass(self, type, version, work_dir):
+        """
+        Returns the Python class that can be used to instantiate a processor of the given type.
+        Additionally, it ensures that the required third-party dependencies are on the system path in order to ensure that
+        the necessary libraries are available to the Processor so that it can be instantiated and used.
+
+        :param type: the type of Processor
+        :param version: the version of the Processor
+        :param work_dir: the working directory for extensions
+        :return: the Python class that can be used to instantiate a Processor of the given type and version
+
+        :raises ValueError: if there is no known Processor with the given type and version
+        """
+        id = ExtensionId(classname=type, version=version)
+        if id in self.processor_class_by_name:
+            return self.processor_class_by_name[id]
+
+        if id not in self.module_files_by_extension_type:
+            raise ValueError('Invalid Processor Type: No module is known to contain Processor of type ' + type + ' version ' + version)
+        module_file = self.module_files_by_extension_type[id]
+
+        if id in self.processor_details:
+            extension_working_dir = os.path.join(work_dir, 'extensions', type, version)
+            sys.path.insert(0, extension_working_dir)
```

Comment:
To answer my own question in my [comment](https://github.com/apache/nifi/pull/7003#issuecomment-1471969300) , `$PYTHONPATH` is extended here. I don't think we should modify the `$PYTHONPATH` when we are already using virtual environments. As I've mentioned, all dependencies should go inside `$VIRTUAL_ENV/lib/$PYTHONVERSION/site-packages'. That way every Python script running in that specific virtual environment can use the dependencies. To achieve this, per processor type virtual environments should be created first, activated, then call `pip install` to install them inside the virtual environment. The current implementation of dependency installation feels kinda backwards to how it's usually done. Is there any technical reason we are doing it this way?

Reply:
> "The current implementation of dependency installation feels kinda backwards to how it's usually done."
The current implementation follows the same steps that you've laid out here. It creates the environment. Then it launches a Process in that environment, which calls pip install. The only difference is really the full "activation" of the environment. While that's easy and straight-forward from commandline, it's less straight-forward programmatically. We do make use of the Python binary that is created for the environment, and we install all requirements/dependencies into that environment.

Again, if you have ideas about how to make this cleaner, I'm happy to review a follow-on PR. Currently, though, this appears to work nicely.